torch/fx/experimental/symbolic_shapes.py

```diff
@@ -186,79 +177,225 @@ def magic_impl(self, other):
             if isinstance(other, PySymInt):
                 other = other.expr
             # TODO: consider constant prop here
-            return PySymInt(func(self.expr, other), self.shape_env)
+            expr = self.shape_env.replace(self.expr)
+            other = self.shape_env.replace(other)
+            out = func(expr, other)
+            out = self.shape_env.replace(out)
+            return PySymInt(out, self.shape_env)
         return magic_impl
 
+    _func = lru_cache(256)(_func)
```

Comment:
🤔 why is this sound?

Reply:
This only operates on `sympy.Expr` calls.